File: Src/Icm.Core.Tests/Collections extensions/ArrayExtensionsTest.vb
Imports System


Imports Icm.Collections



'''<summary>
'''This is a test class for ArrayExtensionsTest and is intended
'''to contain all ArrayExtensionsTest Unit Tests
'''</summary>
<TestFixture(), Category("Icm")>
Public Class ArrayExtensionsTest

    '''<summary>
    '''A test for MultiGetRow
    '''</summary>
    Public Sub MultiGetRowTestString()

        Dim a(,) As String = {{"maria", "rafa"}, {"juan", "pepe"}}
        Dim iteratingDimension As Integer = 0
        Dim fixedDimensionValues() As Integer = Nothing
        Dim expected() As String
        Dim actual() As String

        'Caso 1
        fixedDimensionValues = {1}
        expected = {"rafa", "pepe"}
        actual = ArrayExtensions.MultiGetRow(Of String)(a, iteratingDimension, fixedDimensionValues)
        Assert.IsTrue(expected.SequenceEqual(actual))


    End Sub

    <Test(), Category("Icm")>
    Public Sub MultiGetRow_ExtremeCases()

        Dim a(,) As String
        Dim iteratingDimension As Integer = 0
        Dim fixedDimensionValues() As Integer = Nothing
        Dim expected() As String
        Dim actual() As String

        'Caso 1
        'fixedDimensionValues out of range case
        Try
            a = {{"", ""}, {"", ""}}
            iteratingDimension = 0
            fixedDimensionValues = {2}
            expected = {"", ""}
            actual = ArrayExtensions.MultiGetRow(Of String)(a, iteratingDimension, fixedDimensionValues)
            Assert.Fail("Should not launch when you enter exception index out of range")
        Catch ex As IndexOutOfRangeException
        Catch ex As Exception
            Assert.Fail("Should throw IndexOutOfRangeException")

        End Try

        'Caso 2
        'iteratingDimension es numero negativo
        Try
            a = {{"", ""}, {"", ""}}
            iteratingDimension = -1
            fixedDimensionValues = {1}
            expected = {"", ""}
            actual = ArrayExtensions.MultiGetRow(Of String)(a, iteratingDimension, fixedDimensionValues)
            Assert.Fail("Should not accept negative number")
        Catch ex As IndexOutOfRangeException
        Catch ex As Exception
            Assert.Fail("Should throw IndexOutOfRangeException")
        End Try

        'Caso 3
        'FixedDimensionValues es numero negativo
        Try
            a = {{"", ""}, {"", ""}}
            iteratingDimension = 0
            fixedDimensionValues = {-1}
            expected = {"", ""}
            actual = ArrayExtensions.MultiGetRow(Of String)(a, iteratingDimension, fixedDimensionValues)
            Assert.Fail("Should not accept negative number")
        Catch ex As IndexOutOfRangeException
        Catch ex As Exception
            Assert.Fail("Should throw IndexOutOfRangeException")
        End Try

        'Caso 4
        'FixedDimensionValues out of range case
        Try
            a = {{"", ""}, {"", ""}}
            iteratingDimension = 5
            fixedDimensionValues = {1}
            expected = {"", ""}
            actual = ArrayExtensions.MultiGetRow(Of String)(a, iteratingDimension, fixedDimensionValues)
            Assert.Fail("Should not launch when you enter exception index out of range")
        Catch ex As IndexOutOfRangeException
        Catch ex As Exception
            Assert.Fail("Should throw IndexOutOfRangeException")
        End Try

    End Sub

    '''<summary>
    '''A test for MultiGetRow
    '''</summary>
    Public Sub MultiGetRowTestInteger()

        Dim a(,) As Integer = {{1, 5}, {2, 7}, {3, 9}}
        Dim iteratingDimension As Integer = 1
        Dim fixedDimensionValues() As Integer = Nothing
        Dim expected() As Integer
        Dim actual() As Integer

        'Caso 1
        fixedDimensionValues = {1}
        expected = {2, 7}
        actual = ArrayExtensions.MultiGetRow(Of Integer)(a, iteratingDimension, fixedDimensionValues)
        Assert.IsTrue(expected.SequenceEqual(actual))

    End Sub

    <Test(), Category("Icm")>
    Public Sub MultiGetRowTest_BasicCases()
        MultiGetRowTestString()
        MultiGetRowTestInteger()
    End Sub
End Class

File: Src/Icm.Core.Tests/Collections extensions/ArrayExtensionsTest.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using Icm.Collections;

[TestFixture(), Category("Icm")]
public class ArrayExtensionsTest
{

	static readonly object[] MultiGetRow_NormalStringTestCases = { new object[] {
		new string[ + 1,  + 1] {
			{
				"maria",
				"rafa"
			},
			{
				"juan",
				"pepe"
			}
		},
		0,
		new int[] { 1 },
		new string[] {
			"rafa",
			"pepe"
		}

	} };
	static readonly object[] MultiGetRow_NormalIntegerTestCases = {
		new object[] {
			new int[ + 1,  + 1] {
				{
					1,
					5
				},
				{
					2,
					7
				},
				{
					3,
					9
				}
			},
			1,
			new int[] { 1 },
			new int[] {
				2,
				7
			}
		},
		new object[] {
			new int[ + 1,  + 1,  + 1] {
				{
					{
						11,
						15,
						14
					},
					{
						12,
						17,
						16
					},
					{
						13,
						19,
						18
					}
				},
				{
					{
						21,
						25,
						24
					},
					{
						22,
						27,
						26
					},
					{
						23,
						29,
						28
					}
				},
				{
					{
						31,
						35,
						34
					},
					{
						32,
						37,
						36
					},
					{
						33,
						39,
						38
					}
				}
			},
			0,
			new int[] {
				1,
				1
			},
			new int[] {
				17,
				27,
				37
			}
		},
		new object[] {
			new int[ + 1,  + 1,  + 1] {
				{
					{
						11,
						15,
						14
					},
					{
						12,
						17,
						16
					},
					{
						13,
						19,
						18
					}
				},
				{
					{
						21,
						25,
						24
					},
					{
						22,
						27,
						26
					},
					{
						23,
						29,
						28
					}
				},
				{
					{
						31,
						35,
						34
					},
					{
						32,
						37,
						36
					},
					{
						33,
						39,
						38
					}
				}
			},
			1,
			new int[] {
				1,
				2
			},
			new int[] {
				24,
				26,
				28
			}
		}

	};

	static readonly object[] MultiGetRow_ExceptionalStringTestCases = {
		new object[] {
			new string[ + 1,  + 1] {
				{
					"",
					""
				},
				{
					"",
					""
				}
			},
			0,
			new int[] { 2 },
			typeof(IndexOutOfRangeException)
		},
		new object[] {
			new string[ + 1,  + 1] {
				{
					"",
					""
				},
				{
					"",
					""
				}
			},
			-1,
			new int[] { 1 },
			typeof(IndexOutOfRangeException)
		},
		new object[] {
			new string[ + 1,  + 1] {
				{
					"",
					""
				},
				{
					"",
					""
				}
			},
			0,
			new int[] { -1 },
			typeof(IndexOutOfRangeException)
		},
		new object[] {
			new string[ + 1,  + 1] {
				{
					"",
					""
				},
				{
					"",
					""
				}
			},
			5,
			new int[] { 1 },
			typeof(IndexOutOfRangeException)
		},
		new object[] {
			new string[ + 1,  + 1,  + 1] {
				{
					{
						"",
						""
					},
					{
						"",
						""
					}
				},
				{
					{
						"",
						""
					},
					{
						"",
						""
					}
				}
			},
			0,
			new int[] { 1 },
			typeof(ArgumentException)
		}

	};

	[Test()]
	[TestCaseSource("MultiGetRow_NormalStringTestCases")]
	public void MultiGetRow_ReturnsExpected(string[,] target, int iteratingDimension, int[] fixedDimensionValues, string[] expected)
	{
		dynamic actual = target.MultiGetRow<string>(iteratingDimension, fixedDimensionValues);
		Assert.That(actual, Is.EquivalentTo(expected));
	}

	[Test()]
	[TestCaseSource("MultiGetRow_NormalIntegerTestCases")]
	public void MultiGetRow_ReturnsExpected(Array target, int iteratingDimension, int[] fixedDimensionValues, int[] expected)
	{
		dynamic actual = target.MultiGetRow<int>(iteratingDimension, fixedDimensionValues);
		Assert.That(actual, Is.EquivalentTo(expected));
	}

	[Test()]
	[TestCaseSource("MultiGetRow_ExceptionalStringTestCases")]
	public void MultiGetRow_ThrowsIndexOutOfRange(Array target, int iteratingDimension, int[] fixedDimensionValues, Type exceptionType)
	{
		Assert.That(() => target.MultiGetRow<string>(iteratingDimension, fixedDimensionValues), Throws.TypeOf(exceptionType));
	}

}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
